diff --git a/drivers/gpio/gpio_pcf857x.c b/drivers/gpio/gpio_pcf857x.c
index c30e1a5d36d..361e89e1434 100644
--- a/drivers/gpio/gpio_pcf857x.c
+++ b/drivers/gpio/gpio_pcf857x.c
@@ -135,10 +135,10 @@ static int pcf857x_port_get_raw(const struct device *dev, gpio_port_value_t *val
 		return -EWOULDBLOCK;
 	}
 
-	if ((~drv_data->pins_cfg.configured_as_outputs & (uint16_t)*value) != (uint16_t)*value) {
-		LOG_ERR("Pin(s) is/are configured as output which should be input.");
-		return -EOPNOTSUPP;
-	}
+	// if ((~drv_data->pins_cfg.configured_as_outputs & BIT(*value)) != BIT(*value)) {
+	// 	LOG_ERR("Pin(s) is/are configured as output which should be input.");
+	// 	return -EOPNOTSUPP;
+	// }
 
 	k_sem_take(&drv_data->lock, K_FOREVER);
 
@@ -176,8 +176,7 @@ static int pcf857x_port_set_raw(const struct device *dev, uint16_t mask, uint16_
 	if (k_is_in_isr()) {
 		return -EWOULDBLOCK;
 	}
-
-	if ((drv_data->pins_cfg.configured_as_outputs & value) != value) {
+	if (((drv_data->pins_cfg.configured_as_outputs & value) | ~drv_data->pins_cfg.configured_as_outputs) != (value | ~drv_data->pins_cfg.configured_as_outputs)) {
 		LOG_ERR("Pin(s) is/are configured as input which should be output.");
 		return -EOPNOTSUPP;
 	}
@@ -185,6 +184,7 @@ static int pcf857x_port_set_raw(const struct device *dev, uint16_t mask, uint16_
 	tx_buf = (drv_data->pins_cfg.outputs_state & ~mask);
 	tx_buf |= (value & mask);
 	tx_buf ^= toggle;
+	tx_buf |= ~drv_data->pins_cfg.configured_as_outputs;
 	sys_put_le16(tx_buf, tx_buf_p);
 
 	rc = i2c_write_dt(&drv_cfg->i2c, tx_buf_p, drv_data->num_bytes);
@@ -222,8 +222,10 @@ static int pcf857x_pin_configure(const struct device *dev, gpio_pin_t pin, gpio_
 		return -ENOTSUP;
 	}
 	if (flags & GPIO_INPUT) {
+		drv_data->pins_cfg.configured_as_outputs |= BIT(pin);
+		temp_outputs = drv_data->pins_cfg.configured_as_outputs;
 		temp_outputs &= ~BIT(pin);
-		temp_pins &= ~(1 << pin);
+		temp_pins |= BIT(pin);
 	} else if (flags & GPIO_OUTPUT) {
 		drv_data->pins_cfg.configured_as_outputs |= BIT(pin);
 		temp_outputs = drv_data->pins_cfg.configured_as_outputs;
